internal/compiler/passes/resolving/remove_noop.rs

```diff
@@ -0,0 +1,110 @@
+        // side effects
+        Expression::PropertyReference(_) => false,
+        Expression::ElementReference(_) => false,
+        Expression::RepeaterIndexReference { .. } => true,
+        Expression::RepeaterModelReference { .. } => true,
+        Expression::StoreLocalVariable { .. } => false,
+        Expression::ReadLocalVariable { .. } => true,
+        Expression::StructFieldAccess { base, name: _ } => without_side_effects(&*base),
+        Expression::ArrayIndex { array, index } => {
+            without_side_effects(&*array) && without_side_effects(&*index)
+        }
+        // Note: This assumes that the cast itself does not have any side effects, which may not be
+        // the case if custom casting rules are implemented.
+        Expression::Cast { from, to: _ } => without_side_effects(from),
+        // Note: Calling a *pure* function is without side effects, however
+        // just from the expression, the purity of the function is not known.
+        // We would need to resolve the function to determine its purity.
+        Expression::FunctionCall { .. } => false,
+        Expression::SelfAssignment { .. } => false,
+        Expression::BinaryExpression { lhs, rhs, .. } => {
+            without_side_effects(&*lhs) && without_side_effects(&*rhs)
+        }
+        Expression::UnaryOp { sub, op: _ } => without_side_effects(&*sub),
+        Expression::ImageReference { .. } => true,
+        Expression::Array { element_ty: _, values } => values.iter().all(without_side_effects),
+        Expression::Struct { ty: _, values } => values.values().all(without_side_effects),
+        Expression::PathData(_) => true,
+        Expression::EasingCurve(_) => true,
+        Expression::LinearGradient { angle, stops } => {
+            without_side_effects(&angle)
+                && stops
+                    .iter()
+                    .all(|(start, end)| without_side_effects(start) && without_side_effects(end))
+        }
+        Expression::RadialGradient { stops } => stops
+            .iter()
+            .all(|(start, end)| without_side_effects(start) && without_side_effects(end)),
+        Expression::ConicGradient { stops } => stops
+            .iter()
+            .all(|(start, end)| without_side_effects(start) && without_side_effects(end)),
+        Expression::EnumerationValue(_) => true,
+        // A return statement is never without side effects, as an important "side effect" is that
+        // the current function stops at this point.
+        Expression::ReturnStatement(_) => false,
+        Expression::LayoutCacheAccess { .. } => false,
+        Expression::ComputeLayoutInfo(_, _) => false,
+        Expression::SolveLayout(_, _) => false,
+        Expression::MinMax { ty: _, op: _, lhs, rhs } => {
+            without_side_effects(lhs) && without_side_effects(rhs)
+        }
+        Expression::DebugHook { .. } => false,
```

Comment:
for the purpose of the warning, it shouldn't have side effect

Reply:
If it was without side effects (`true`), this means it could be removed.
I don't know the details of how DebugHooks work inside of Slint, but it sounds dangerous to remove. :thinking: